### PR TITLE
CCM-7462: add login logout pages on dev server

### PR DESCRIPTION
**Pull request by harrim91**
## Description

Adds sign in and sign out pages to templates app which are only available when running the dev server.
Makes whole app auth-aware by wrapping in Amplify Auth Provider.

## Context

Allows us to authenticate when running locally without the auth web app available.

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: next.config.js [hunk 1]
```diff
@@ -24,6 +27,12 @@ const nextConfig = {
       },
     ];
   },
-};
+
+  // pages with e.g. .dev.tsx extension are only included when running locally
+  pageExtensions: ['ts', 'tsx', 'js', 'jsx'].flatMap((extension) => {
```
Comment (chris-elliott-nhsd) [inline next.config.js:32]: this does not work because .dev.tsx extensions are still .tsx extensions, so the dev pages will still get deployed everywhere

Comment (harrim91) [reply next.config.js]: It works for me. I deleted `frontend/.next` directory, ran `npm run build` to create a production build and then `npm start` to run the production build. The login and logout pages weren't available and I got a 404 instead.

Comment (chris-elliott-nhsd) [reply next.config.js]: wow. my mistake! i feel like i tried to do this before and ended up having to rename all the other pages but if this works fine

File: frontend/amplify/package.json [hunk 2]
```diff
@@ -4,6 +4,14 @@
   "private": true,
   "main": "index.ts",
   "type": "module",
+  "scripts": {
+    "generate-dependencies": "tsx functions/send-email/generate-dependencies.ts",
+    "lint": "eslint . -c .eslintrc.json",
+    "lint:fix": "eslint . --fix",
+    "test:unit": "jest",
+    "typecheck": "tsc --noEmit",
+    "typescript": "^5.5.4"
```
Comment (bhansell1) [inline frontend/amplify/package.json:13]: How comes this is in scripts and not dependencies?

File: frontend/src/app/auth/page.dev.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
+'use client';
+
+import React, { Suspense } from 'react';
+import { useSearchParams, redirect, RedirectType } from 'next/navigation';
+import { Authenticator } from '@aws-amplify/ui-react';
+import '@aws-amplify/ui-react/styles.css';
+// this is an npm module, not node:path
+// eslint-disable-next-line  unicorn/prefer-node-protocol
+import path from 'path';
```
Comment (bhansell1) [inline frontend/src/app/auth/page.dev.tsx:9]: path is in node. Why the change? 

https://nodejs.org/docs/latest-v18.x/api/path.html#pathnormalizepath

File: lambdas/endpoint/dist/index.js
Comment (bhansell1) [inline lambdas/endpoint/dist/index.js:1]: I don't think endpoint should still be here?

File: frontend/amplify/package.json (hunk 2 above)
Comment (harrim91) [reply frontend/amplify/package.json:13]: 🤔

Comment (harrim91) [reply frontend/amplify/package.json:13]: Good spot. It was already in `scripts` and I just moved the entire `scripts` block. But I will fix it.

File: frontend/src/app/auth/page.dev.tsx (hunk 3 above)
Comment (harrim91) [reply frontend/src/app/auth/page.dev.tsx:9]: This `Redirect` component was copied from the auth webapp - https://github.com/NHSDigital/nhs-notify-iam-webauth/blob/main/src/components/molecules/Redirect/Redirect.tsx

I'm guessing an npm package is being used because it's a client component so doesn't have access to Node core modules.

File: lambdas/endpoint/dist/index.js
Comment (harrim91) [reply lambdas/endpoint/dist/index.js:1]: 😳

File: frontend/amplify/package.json (hunk 2 above)
Comment (harrim91) [reply frontend/amplify/package.json]: fixed

File: frontend/src/app/auth/page.dev.tsx (hunk 3 above)
Comment (bhansell1) [reply frontend/src/app/auth/page.dev.tsx:9]: ahhhhh fair!

**Review by chris-elliott-nhsd — APPROVED** [review]

File: lambdas/backend-api/package.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
   "version": "0.0.1",
   "scripts": {
     "generate-dependencies": "tsx scripts/generate-dependencies.ts",
-    "test:unit": "jest",
+    "test:unit": "npm run generate-dependencies && jest",
```
Comment (chris-elliott-nhsd) [inline lambdas/backend-api/package.json:7]: if we're running generate dependencies as part of the test:unit commands, we don't need to run it separately on the pipeline right? no strong views which one we keep but feels like we could save time by not having both

**Review by bhansell1 — APPROVED** [review]

**Review by chris-elliott-nhsd — APPROVED** [review]

File: lambdas/backend-api/package.json (hunk 4 above)
Comment (harrim91) [reply lambdas/backend-api/package.json:7]: I added it here because I ran the test script locally and it failed because I hadn't run the script. I think it was the first time I'd run it since the workspaces thing came in and the file structure was changed. I added it because you need the dependency to make sure the tests pass, so it made sense to make the test command run it. 

But I've just realised that the linting and typecheck also fail if you don't have the `email-template.json` file. So with this approach, really I should be prefixing all of those commands with `generate-dependencies`. Which does seem a bit pointless - it'd be better to stick with the current approach and run it once manually when needed. 

And also there are a few packages that have the `generate-dependencies` script, and the pipelines are running it in all the workspaces. Whereas this is just in the `backend-api` package. So we'd still need to keep it in the pipeline for those other packages possibly.

So in conclusion, I will remove this!

Comment (harrim91) [reply lambdas/backend-api/package.json]: Updated

**Review by chris-elliott-nhsd — APPROVED** [review]
